# Polygenic Risk Score Estimation with LDpred and PLINK

## Requirements
This procedure requires LDpred and PLINK (Version 1 *and* 2) to be installed.
Furthermore, the following R packages should be installed:

* data.table
* magrittr
* stringr

They are available in the `R_packages` module on UPPMAX.

All scripts are provided as separate Slurm and R files but they need some customization for your specific use-case.
I make heavy use of Slurm Job Arrays to parallelize the operations.
If you aren't familiar with them, the [documentation on the slurm website](https://slurm.schedmd.com/job_array.html) is pretty good.


## Directory Structure
This manual and all scripts assume the following directory structure:

```
.
├── coordination
├── genotypes
├── gwas
├── ld
├── mfi
├── reference_genotypes
├── sample_info
├── scores
└── weights
```

Where all directories are *empty*, except for:
* `genotypes` containing all your plink genotype/SNP data
* `sample_info` for your fam files,
* `mfi` containing MAF and imputation quality information

You can use [`make_dirs`](make_dirs) to create all the empty directories in one step.
I recommend creating the non-empty directories as symlinks to the shared locations holding all your required data.

## Preparing the LD Estimation

### Formatting your GWAS Summary
LDpred requires a GWAS summary.
It accepts a range of formats.
However, all of them are different from the one generated by Plink.
We use the file format which was the default in the LDpred beta releases, called `STANDARD`.

It is as follows:

```
chr pos ref alt reffrq info rs pval effalt
```

Where:
* `chr` is the SNP's chromosome
* `pos` is the SNP's base pair
* `ref` is the reference allele
* `alt` is the alternative allele
* `reffrq` is the frequency of the **reference** allele
* `info` is the imputation quality as info score
* `rs` is the rs ID
* `pval` is the SNP's P value
* `effalt` is the beta value

*Note*: If you use any version of LDpred before 1.0.0, the info score is ignored and you can safely set it to 1.

You can find all the necessary steps in [`LDpred_formatting.R`](LDpred_formatting.R).

While you are at it: This is the perfect moment to filter for imputation quality.
Subset your data by INFO score and save the passing rs IDs to a file (referred to as `info_snps.txt` below).

### Extracting the Reference Dataset
#### Subsetting Individuals
LDpred uses a reference dataset in order to compute the LD pattern.
A good option is to use 5000 random non-related individuals.
You can extract this kind of set with the following command:

```
shuf -n 5000 sample_info/samples.fam > 5000_individuals.txt
```

LDpred works on Plink 1 binary (i.e. bed/bim/fam) files.
Therefore, we need to extract the individuals' genotypes and save them to BED files.

Please see [`extract.slurm`](extract.slurm) for the extraction script.

#### Merging the Genotyping Data
The resulting bed files must be merged into one.
Unfortunately, Plink 2 doesn't support merging, yet, so we must do it with Plink 1.
Once merging is implemented in version 2, this step can be combined with the one above into one step.

First, you must create a list of all data sets to be merged.
You can use [`create_merge_list.R`](create_merge_list.R) for that.

Please see [`merge.slurm`](merge.slurm) for the merging script.

## Genotype Coordination
The first step of your LDpred analysis is the coordination of your genotyping data and GWAS summary.

Please see [`coord.slurm`](coord.slurm) for the appropriate script.
Before submitting a job you *must* adjust parameter `N`, which is the number of individuals in your GWAS.

**Important:** If the coordination fails and you want to restart it, remove the file `coordination/coordination.coord` first.
Otherwise the job will fail, again.

## Estimating SNP weights
After the coordination is done, you can estimate SNP weights.
Please see [`estimate_weights.slurm`](estimate_weights.slurm) for a template.
The call has two parameters which need adjustment:
* `N`: The number of individuals in your GWAS (analogous to the coordination step)
* `ldr`: The LD radius, i.e. the number of SNPs around the focal SNP which LD should be adjusted for. Recommended value: # of total SNPs / 3000

The optional parameter `f` is used to specify the fraction of causal variants assumed by LDpred.
It defaults to `1 0.1 0.01 0.001 0.3 0.03 0.003 inf`.
There are many additional arguments you can tinker with.
It's best to check the documentation using
```
ldpred gibbs --help
```

The LD pattern is saved to the file given as the parameter `ldf`.
This file can be reused across analyses which use the same LD radius.
If it doesn't exist, it will be created, which may take several hours.

## Calculating Your Scores
After LDpred has finished estimation LD and SNP weights, it's time to finally compute polygenic risk scores.
I did it with plink but this functionality is offered by LDpred, as well.